### PR TITLE
Add Doc: Customize S2I Templates

### PR DESCRIPTION
**Pull request by Felixnoo**
Signed-off-by: Felixnoo <felixliu@yunify.com>

### CODE REVIEW

**Comment by Felixnoo** [comment]
/assign @Sherlock113 
/cc @kubesphere/sig-devops 
Please review the new doc I added, thanks.

**Comment by Felixnoo** [comment]
fix #1381

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 1]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+  - save-artifacts (optional): the `save-artifacts` script that manages all dependencies in an incremental building process.
+  - usage (optional): the script that provides instructions.
+  - test (optional): the script for testing.
+- Part 2: definition of S2I template
+
+You need to have the required elements for S2I template customization ready in advance.
+
+{{< notice note >}}
+
+The Image Builder is compatible with that of OpenShift, and you can reuse it in KubeSphere. For more information about S2I Image Builder, refer to [S2IRun](https://github.com/kubesphere/s2irun/blob/master/docs/builder_image.md#s2i-builder-image-requirements).
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:29]: S2I Image Builder
->
S2I Image Builders

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 2]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+
+{{</ notice >}}
+
+## Create an Image Builder
+
+### Step 1: Prepare S2I directory
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:35]: Prepare S2I directory
->
Prepare an S2I directory

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 3]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+
+1. [S2I command line tool](https://github.com/openshift/source-to-image/releases) provides an easy-to-use command to initialize a base directory structure required by the Builder. Run the following commands to install S2I CLI.
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:37]: [S2I command line tool](https://github.com/openshift/source-to-image/releases) provides
->
The [S2I command line tool](https://github.com/openshift/source-to-image/releases) provides

S2I CLI
->
the S2I CLI

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 4]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+Modify the Dockerfile as follows to define the Image Builder.
+
+#### Dockerfile
+
+```
+# nginx-centos7
+FROM kubespheredev/s2i-base-centos7:1
+
+# Here you can specify the maintainer for the image that you're building
+LABEL maintainer="Runze Xia <runzexia@yunify.com>"
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:81]: Use an example name and email address, such as `maintainer name` and `email@xxx.com`.

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 5]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+
+{{< notice note >}}
+
+S2I scripts will use the flags defined in the Dockerfile as parameters. If you need to use a base image different from those provided by KubeSphere, refer to [S2I Scripts](https://github.com/kubesphere/s2irun/blob/master/docs/builder_image.md#s2i-scripts).
+
+{{</ notice >}}
+
+### Step 3: Create S2I Scripts
+
+1. Create an `assemble` script as follows to copy configuration file and static contents to the target container.
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:128]: copy configuration file
->
copy the configuration file

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 6]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+   	docker build -t $(IMAGE_NAME) .
+   
+   # The Image Builder can be tested using the following commands:
+   .PHONY: test
+   test:
+   	docker build -t $(IMAGE_NAME)-candidate .
+   	IMAGE_NAME=$(IMAGE_NAME)-candidate test/run
+   ```
+
+2. Run the `make build` command to build the Image Builder for Nginx.
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:201]: Nginx
->
NGINX

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 7]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+
+   ```bash
+   docker run -p 8080:8080  sample-app
+   ```
+
+   You can access the Nginx application at `http://localhost:8080`.
+
+   ![access-nginx](/images/docs/project-user-guide/image-builder/s2i-templates/access-nginx.png)
+
+### Step 5: Push image and create S2I template
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:257]: Push image and create S2I template
->
Push an image and create an S2I template

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 8]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+metadata:
+  labels:
+    controller-tools.k8s.io: "1.0"
+    builder-type.kubesphere.io/s2i: "s2i"
+  name: nginx-demo
+spec:
+  containerInfo:
+    - builderImage: kubespheredev/nginx-centos7-s2ibuilder-sample
+  codeFramework: nginx # type of code framework
+  defaultBaseImage: kubespheredev/nginx-centos7-s2ibuilder-sample # default Image Builder (can be replaced by customized image)
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:275]: by customized image
->
by a customized image

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 9]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+  version: 0.0.1 # Builder template version
+  description: "This is a S2I builder template for Nginx builds whose result can be run directly without any further application server.." # Builder template description
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:277]: This is a S2I builder template for Nginx builds whose result can be run directly without any further application server..
->
This is an S2I builder template for NGINX builds whose result can be run directly without any further application server.

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 10]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+```
+
+### Step 6: Use S2I template on KubeSphere
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:280]: Use S2I template
->
Use the S2I template

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 11]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+
+1. Run the following command to submit the S2I template created above to KubeSphere.
+
+   ```bash
+   $ kubectl apply -f s2ibuildertemplate.yaml
+   s2ibuildertemplate.devops.kubesphere.io/nginx created
+   ```
+
+2. You can find the customized S2I template available when you create a S2I build on KubeSphere.
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:289]: a S2I build
->
an S2I build

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 12]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+
+   ![template-available](/images/docs/project-user-guide/image-builder/s2i-templates/template-available.png)
+
+## S2I Template Parameters Definition
+
+Refer to the following detailed descriptions of S2I template labels passed as parameters to frontend classifications.
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:295]: to frontend classifications.
->
for frontend classification.

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 13]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+
+| Label Name                            | Option               | Definition                                                   |
+| ------------------------------------- | -------------------- | ------------------------------------------------------------ |
+| builder-type.kubesphere.io/s2i: "s2i" | "s2i"                | The type of this template is S2I, which builds images based on application source code. |
+| builder-type.kubesphere.io/b2i        | "b2i"                | The type of this template is B2I, which builds images based on binary files or other artifacts. |
+| binary-type.kubesphere.io             | "jar","war","binary" | This type is complementary to the type of B2I and will be required when B2I is selected. For example, select the type of "jar" when a JAR package is provided. In KubeSphere v2.1.1 and later, it is also allowed to customize B2I template. |
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:301]: "jar","war","binary"
->
"jar", "war", "binary"

customize B2I template
->
customize B2I templates

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 14]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+
+Refer to the following detailed descriptions of S2I template parameters. The required parameters are marked with an asterisk.
+
+| Parameter                                  | Type     | Definition                                                   |
+| ------------------------------------------ | -------- | ------------------------------------------------------------ |
+| *containerInfo                             | []struct | The information about Image Builder.                         |
+| *containerInfo.builderImage                | string   | S2I Image Builder, such as kubesphere/java-8-centos7:v2.1.0. |
+| containerInfo.runtimeImage                 | string   | S2I Runtime Image, such as kubesphere/java-8-runtime:v2.1.0. |
+| containerInfo.buildVolumes                 | []string | The information about mounted volume. The format is "volume_name:mount_path", such as ["s2i_java_cache:/tmp/artifacts","test_cache:test_path"]. |
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:310]: about mounted volume
->
about mounted volumes

File: content/en/docs/project-user-guide/image-builder/s2i-templates.md [hunk 15]
```diff
@@ -6,4 +6,323 @@ linkTitle: "Customize S2I Templates"
...
+| containerInfo.runtimeArtifacts             | []struct | The list of original path and target path for the output artifact; only add it for phased building. |
+| containerInfo.runtimeArtifacts.source      | string   | The original path of artifact in Image Builder.              |
+| containerInfo.runtimeArtifacts.destination | string   | The target path of artifact in Runtime Image.                |
+| containerInfo.runtimeArtifacts.keep        | bool     | Whether to keep the data in the output image.                |
+| *defaultBaseImage                          | string   | The default Image Builder.                                   |
+| *codeFramework                             | string   | The code framework type, such as Java, Ruby.                 |
```
Comment (Sherlock113) [inline content/en/docs/project-user-guide/image-builder/s2i-templates.md:316]: Java, Ruby
->
Java and Ruby

**Pull request merged**